packages/typespec-client-generator-core/src/internal-utils.ts

```diff
@@ -252,12 +255,85 @@ export function getSdkTypeBaseHelper<TKind>(
   context: TCGCContext,
   type: Type,
   kind: TKind
-): DefaultSdkTypeBase<TKind> {
-  return {
+): [DefaultSdkTypeBase<TKind>, readonly Diagnostic[]] {
+  const diagnostics = createDiagnosticCollector();
+  return diagnostics.wrap({
     __raw: type,
     deprecation: getDeprecationDetails(context.program, type)?.message,
     kind,
-  };
+    decorators: diagnostics.pipe(getTypeDecorators(context, type)),
+  });
+}
+
+export function getTypeDecorators(
+  context: TCGCContext,
+  type: Type
+): [Record<string, Record<string, any>>, readonly Diagnostic[]] {
+  const diagnostics = createDiagnosticCollector();
+  const retval: Record<string, Record<string, any>> = {};
+  if ("decorators" in type) {
+    for (const decorator of type.decorators) {
+      // only process explicitly defined decorators
+      if (decorator.definition) {
+        const decoratorName = `${decorator.definition?.namespace ? getNamespaceFullName(decorator.definition.namespace) + "." : ""}${decorator.definition?.name}`;
+        // white list filtering
+        if (
+          !context.decoratorsWhiteList ||
+          !context.decoratorsWhiteList.some(
+            (x) => x === decoratorName || new RegExp(x).test(decoratorName)
```

Comment:
so with this, if I want to specify one certain decorator `foo` in the configuration, I will have to write this:
```
^foo$
```
because `foo` might match other names such as `barfoobaz`
Would this happen? And is this expected?

Reply:
Also based on your test cases, and since dot `.` is quite common in the name of decorators, if I write this: `TypeSpec.@bar`, would it leads to a regex that matches anything starting with `TypeSpec` and ending with `@bar`? This might match all the decorators with the name `@bar` in all the sibling namespaces under `TypeSpec`, also `TypeSpec2.@bar` maybe
if this is true, this really introduces quite a lot uncertainties

Reply:
you are right. the two comparisons have overlap. let me change to only allow regex.